[lld][MachO] Follow-up to use madvise() for threaded file page-in.

johnno1962 opened the pull request:
Further to https://github.com/llvm/llvm-project/pull/147134#discussion_r2337246489, switch to use the madvise() api to page in mmap'd files and

1) All new code compiled in #if LLVM_ENABLE_THREADS is set so it can be seen where the changes were from this PR.
2) The new PR moves to use madvise() instead of the ad-hoc page referencing code I wrote which should avoid SIGSEGVs if the buffer is deallocated.
3) A new property SerialBackgroundQueue().stopAllWork to be used to stop background workers when there is no further call for them. Usually the background "page-in" threads have completed first but it seems with this troublesome test this is not always the case and buffers stored in the static input file cache are being deallocated while being referenced.

johnno1962 committed:
Switch to use madvise() to page-in files.

aganea commented on lld/MachO/Driver.cpp:
```diff
@@ -349,13 +352,19 @@ void multiThreadedPageInBackground(DeferredFiles &deferred) {
+    numDeferedFilesAdvised += 1;
 #endif
 
+#if _WIN32
```
I would put all this in a API declared in `llvm/include/llvm/Support/Memory.h` and defined in `llvm/lib/Support/Windows/Memory.inc` for Windows and `llvm/lib/Support/Unix/Memory.inc` for Unix/MacOS. Perhaps something like `Memory::prefetchMappedMemory` to be in line with the other APIs there.

johnno1962 replied:
I agree but I'm sorry but I'm haven't signed on for larger refactors.

aganea reviewed (approved):
LGTM.

johnno1962 committed:
Response to review.

johnno1962 committed:
Use LLVM_DEBUG.

johnno1962 committed:
Page align buffer pointer and address 2 second performance regression introduced by:
https://github.com/llvm/llvm-project/commit/85cd3d98686c47d015dbcc17f1f7d0714b00e172

johnno1962 committed:
Alternative solution to performance regression.

johnno1962 committed:
Revert "Alternative solution to performance regression."

This reverts commit 884b771e18b305b6d15b789dbb40d8025798e7cc.

johnno1962 committed:
Format llvm/lib/Object/Archive.cpp

Co-authored-by: James Henderson <James.Henderson@sony.com>

johnno1962 committed:
Format lld/MachO/InputFiles.cpp

johnno1962 committed:
A third alternative.

johnno1962 committed:
Safer IsText option.

johnno1962 committed:
Revert "Safer IsText option."

This reverts commit 198d9b08934af01f19a6fbe92e0d4a5928832d54.

johnno1962 committed:
Revert "A third alternative."

This reverts commit f2654ae6e1bb7aa78a16badb1bcc9d6f1f824676.

johnno1962 committed:
Multi-threaded file opening/mapping.

johnno1962 commented:
Another more speculative commit, multi-threading the opening of files by making macho::readFile() thread safe. Another 1.5 second reduction of time spent linking my example Chrome task.

Thu Sep 18 11:07:02 be7af98b9a94bdb40a918130189544c8ab5232cd
0m10.548s 0m10.581s 0m10.748s 0m10.485s 0m10.486s 0m10.619s 0m10.434s 0m10.682s

johnno1962 committed:
Revert "Multi-threaded file opening/mapping."

This reverts commit be7af98b9a94bdb40a918130189544c8ab5232cd.

johnno1962 committed:
Merge branch 'main' into threaded-advising

johnno1962 committed:
Pre-empt read-worker threads.

aganea reviewed (commented):
As a general comment for this PR, we avoid usually `#ifdefs` in code. I think you can write this entire PR without ever using any `#ifdef LLVM_ENABLE_THREADS`; either by leaving the code active at all times, or by using existing LLVM APIs that already handle that flag. An example of this is `llvm::get_threadpool_strategy` which falls back to not using threads when `LLVM_ENABLE_THREADS` is set. Similarly, `ThreadPool` does the same thing. Also, all places where we test for `if (config->readWorkers)` we could also test for `llvm_is_multithreaded()`.

johnno1962 committed:
Remove #ifdefs.

johnno1962 commented on lld/MachO/Driver.cpp:
```diff
@@ -330,61 +336,65 @@ class SerialBackgroundQueue {
-#ifndef NDEBUG
+
   auto dt = high_resolution_clock::now() - t0;
   if (Process::GetEnv("LLD_MULTI_THREAD_PAGE"))
     llvm::dbgs() << "multiThreadedPageIn " << totalBytes << "/"
-                 << numDeferedFilesTouched << "/" << deferred.size() << "/"
+                 << numDeferedFilesAdvised << "/" << deferred.size() << "/"
                  << duration_cast<milliseconds>(dt).count() / 1000. << "\n";
```
Not sure reading a single value from the environment is that expensive an operation but I've put the NDEBUGs back in. I took them out as function was getting pretty messy and it would likely be useful to have the option to print those numbers when investigating performance of a release build.

DavidSpickett commented:
This would fix an issue on our buildbot, any chance of getting another round of review so we can eventually merge this?

johnno1962 committed:
Merge branch 'main' into threaded-advising

johnno1962 committed:
Put #if NDEBUGs back in.

drodriguez reviewed (changes requested):
This PR doesn't change the tests to not be `UNSUPPORTED` everywhere: https://github.com/llvm/llvm-project/blob/c98e867822d8e741159ff0be09ccc2a5c7841a28/lld/test/MachO/read-workers.s#L2-L4

This is not being tested at all in CI. Should that be changed?

johnno1962 committed:
Re-enable test.

johnno1962 commented:
🎉 Thanks! Let me know if the tests are still fai.

DavidSpickett commented:
The Arm 32-bit problems certainly have not returned.
